pricegraph/src/graph/subgraph.rs

```diff
@@ -1,44 +1,39 @@
 //! Module implementing tools for iterating over disconnected subgraphs.
 
-use petgraph::graph::NodeIndex;
+use super::shortest_paths::ShortestPathGraph;
+use petgraph::visit::{Data, GraphBase, IntoEdges, NodeIndexable};
+use std::cmp::Ord;
 use std::collections::BTreeSet;
 
 /// A struct used for iterating over disconnected subgraphs in the orderbook for
 /// detecting orderbook overlaps and reducing the orderbook.
 ///
 /// Note that this pseudo-iterator uses a `BTreeSet` to ensure that subgraphs
 /// are visited in a predictable order starting the from the first node.
-pub struct Subgraphs(BTreeSet<NodeIndex>);
-
-impl Subgraphs {
-    /// Create a new subgraphs iterator from an iterator of node indices.
-    pub fn new(nodes: impl Iterator<Item = NodeIndex>) -> Self {
-        Subgraphs(nodes.collect())
-    }
-
-    /// Iterate through each subgraph with the provided closure returning the
-    /// predecessor vector for the current node indicating which nodes are
-    /// connected to it.
-    pub fn for_each(self, mut f: impl FnMut(NodeIndex) -> Vec<Option<NodeIndex>>) {
-        self.for_each_until(|node| <ControlFlow<()>>::Continue(f(node)));
+pub struct Subgraphs<G: GraphBase>(BTreeSet<G::NodeId>);
```

Comment:
I think this can be kept as is if we return `shortest_path_graph.connected_nodes()` instead of the `ShortestPathGraph` so that the `Subgraphs` does not need to have a lifetime tied to a graph reference.